src/Microsoft.Diagnostics.Runtime/src/Linux/LinuxLiveDataReader.cs

```diff
@@ -0,0 +1,517 @@
+                    }
+                    result.Add(moduleInfo);
+                }
+            }
+            return result;
+        }
+
+        public void GetVersionInfo(ulong addr, out VersionInfo version)
+        {
+            foreach (var entry in _memoryMapEntries)
+            {
+                if (addr >= entry.BeginAddr && addr <= entry.EndAddr && !string.IsNullOrEmpty(entry.FilePath))
+                {
+                    Version v = null;
+                    int i1 = entry.FileName.LastIndexOf(".so.");
+                    if (i1 > 0)
+                    {
+                        v = this.ParseForVersion(entry.FileName.Substring(i1 + 4));
+                    }
+                    if (v == null)
+                    {
+                        string dirName = Path.GetFileName(Path.GetDirectoryName(entry.FilePath));
+                        v = this.ParseForVersion(dirName);
+                    }
+                    if (v != null)
+                    {
+                        version = new VersionInfo(v.Major, v.Minor, v.Build, v.Revision);
+                        return;
+                    }
+                }
+            }
+            version = new VersionInfo();
+        }
+
+        public bool ReadMemory(ulong address, byte[] buffer, int bytesRequested, out int bytesRead)
+        {
+            this.OpenMemFile();
+            bytesRead = 0;
+            int readableBytesCount = this.GetReadableBytesCount(address, bytesRequested);
+            if (readableBytesCount <= 0)
+            {
+                return false;
+            }
+            try
+            {
+                _memoryStream.Seek((long)address, SeekOrigin.Begin);
+                bytesRead = _memoryStream.Read(buffer, 0, readableBytesCount);
+                return bytesRead > 0;
+            }
+            catch (Exception)
+            {
+                return false;
+            }
+        }
+
+        public bool ReadMemory(ulong address, IntPtr buffer, int bytesRequested, out int bytesRead)
+        {
+            this.OpenMemFile();
+            bytesRead = 0;
+            int readableBytesCount = this.GetReadableBytesCount(address, bytesRequested);
+            if (readableBytesCount <= 0)
+            {
+                return false;
+            }
+            try
+            {
+                byte[] bytes = new byte[readableBytesCount];
+                _memoryStream.Seek((long)address, SeekOrigin.Begin);
+                bytesRead = _memoryStream.Read(bytes, 0, readableBytesCount);
+                if (bytesRead > 0)
+                {
+                    Marshal.Copy(bytes, 0, buffer, bytesRead);
+                }
+                return bytesRead > 0;
+            }
+            catch (Exception)
+            {
+                return false;
+            }
+        }
+
+        public ulong ReadPointerUnsafe(ulong address)
+        {
+            byte[] ptrBuffer = new byte[IntPtr.Size];
```

Comment:
Please make this an instance field on the class.  That way we don't constantly allocate when reading pointers.

Reply:
sure